cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,335 @@
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		isRunning, err := isFauxRunIdRunningYet(config, runsAddr, runId.String())
+		if err != nil {
+			return resource.RetryableError(err)
+		}
+		if isRunning {
+			d.SetId(runId.String())
+			return nil
+		}
+		time.Sleep(10 * time.Second) // avoid spamming retries
+		return resource.RetryableError(fmt.Errorf("still waiting for program run with faux run id: %s which is in an initializing state", runId.String()))
+	})
+}
+
+//TODO ?
+func resourceStreamingProgramRunRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+type RuntimeArgs struct {
+	FauxRunId string `json:"__FAUX_RUN_ID__"`
+}
+
+type RuntimeProperties struct {
+	RuntimeArgs json.RawMessage `json:"runtimeArgs"`
+}
+
+// These are the only keys we need
+type Run struct {
+	RunId      string `json:"runid"`
+	Status     string `json:"status"`
+	Properties RuntimeProperties
+}
+
+// Checks if there is a running run for the terraform faux run id
+// raises error if the program is not in an initializing state (e.g. it failed or was killed in the ui)
+func isFauxRunIdRunningYet(config *Config, runsAddr string, runId string) (bool, error) {
+	s, err := getProgramRunStatusByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return false, err
+	}
+
+	if s == "RUNNING" {
+		return true, nil
+	}
+	if !INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+		return false, fmt.Errorf("Program not Running or Initializing, in state: %s", s)
+	}
+	return false, nil
+
+}
+
+func getProgramRunStatusByFauxId(config *Config, runsAddr string, runId string) (s string, err error) {
+	r, err := getRunByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return
+	}
+
+	return r.Status, nil
+}
+
+// TODO optimization: we call this function often when we could probably get the real runid once and cache it.
+// This would avoid redoing this loop everytime to get the same result. probably inconsequential unless there are many runs of this program
+func getRunByFauxId(config *Config, runsAddr string, runId string) (r Run, err error) {
+	req, err := http.NewRequest(http.MethodGet, runsAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	var runs []Run
+	var args RuntimeArgs
+
+	err = json.Unmarshal(b, &runs)
+	if err != nil {
+		return
+	}
+
+	for _, r := range runs {
+		// Unescaping runtime Args which are stored as an escaped JSON string.
+		unquotedArgs, _ := strconv.Unquote(string(r.Properties.RuntimeArgs))
+		log.Println(fmt.Sprintf("%s", unquotedArgs))
+		b := []byte(unquotedArgs)
+		if err := json.Unmarshal(b, &args); err != nil {
+			// This happens when a run does not contain the special TF con
+			log.Println("failed to parse unquotedArgs json")
```

Comment:
This doesn't seem right.. If there is no faux ID then shouldn't runtimeargs be empty and thus you should get an empty string? And if you unmarshal that then it should just go to a struct that has an empty FauxRunID value.

Reply:
Not necessarily.
If someone triggered a program run in the UI (or by some other deployment process for batch pipelines) there will be no faux id.

Reply:
I don't think you will get an error if the faux id is missing: https://play.golang.org/p/V61GrBnLgHt

So if this returns an error there is probably something more sinister that's wrong, can you check returning error here and seeing if it works?